staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -338,6 +362,28 @@ func New(opts Options) (*Authenticator, error) {
 	return authenticator, nil
 }
 
+// discoveryURLRoundTripper is a http.RoundTripper that rewrites the issuer URL to the
+// discovery URL for the /.well-known/openid-configuration endpoint.
+type discoveryURLRoundTripper struct {
+	base         http.RoundTripper
+	discoveryURL *url.URL
+}
+
+func (t *discoveryURLRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// If the discovery URL is set in authentication configuration, we should
+	// use the discovery URL to fetch the openid configuration. This is useful
+	// for self-hosted providers, for example, providers that run on top of
+	// Kubernetes itself.
+	if req.Method == http.MethodGet && strings.HasSuffix(req.URL.Path, wellKnownEndpointPath) {
+		clone := req.Clone(req.Context())
+		clone.URL.Host = t.discoveryURL.Host
+		clone.URL.Path = t.discoveryURL.Path + wellKnownEndpointPath
+		clone.URL.RawPath = ""
```

Comment:
why not override the whole request URL to discoveryURL, and require discoveryURL to include the well-known path suffix?

as-is, you couldn't use discoveryURL to point to a location for the issuer info that didn't end in a `/.well-known/openid-configuration` suffix. That seems useful to be able to do (like to host issuer info out of a bucket or something that doesn't have that directory structure)

Reply:
With this implementation, the behavior is same as using `issuer.url` for discovery; adding `/.well-known/openid-configuration` during runtime. The feature to [override discovered issuer URL in go-oidc](https://github.com/coreos/go-oidc/pull/315) was implemented similarly (_only overriding the host_), so I was leaning more towards what's generally used.

Reply:
I don't have a strong opinion here, but the current implementation matches what is possible with the "insecure issuer URL" function of go-oidc v3.  If we care to be able to collapse onto that functionality any time in the future, then we need to handle appending the well known suffix.  But this round tripper is minimal enough that I am not worried about keeping it forever.

Reply:
honestly, I find this roundtripper translation easier to reason about than creating an InsecureIssuerURLContext and wondering what is in the context and what it is used for